File: 백준/Gold/2613. 숫자구슬/숫자구슬.py
def check(arr, max_sum):
    """주어진 max_sum으로 나눌 때 필요한 그룹 개수를 계산"""
    cnt = 1
    group_sum = 0
    for num in arr:
        if group_sum + num > max_sum:
            cnt += 1
            group_sum = num
        else:
            group_sum += num
    return cnt

def divide_groups(arr, max_sum, M):
    """M개의 그룹을 정확히 만들어서 그룹 크기를 반환"""
    groups = []
    current_sum = 0
    current_count = 0
    extra_groups = check(arr, max_sum) - M  # 초과된 그룹 개수 (줄여야 할 개수)

    for num in arr:
        if current_sum + num > max_sum or (extra_groups > 0 and len(groups) + (len(arr) - sum(groups)) > M):
            groups.append(current_count)
            current_sum = num
            current_count = 1
            extra_groups -= 1  # 초과 그룹을 줄여야 하므로 감소
        else:
            current_sum += num
            current_count += 1

    groups.append(current_count)

    # 정확히 M개가 될 수 있도록 조정
    while len(groups) < M:
        for i in range(len(groups)):
            if groups[i] > 1:
                groups[i] -= 1
                groups.insert(i + 1, 1)
                if len(groups) == M:
                    break

    return groups

def solve():
    # 입력 받기
    global N, M
    N, M = map(int, input().split())
    arr = list(map(int, input().split()))

    # 이분 탐색 범위 설정
    low, high = max(arr), sum(arr)
    result = high  # 최적의 최대 그룹 합

    while low <= high:
        mid = (low + high) // 2
        if check(arr, mid) <= M:
            result = mid
            high = mid - 1
        else:
            low = mid + 1

    # 최적의 result 값을 기반으로 정확히 M개의 그룹을 만듦
    best_groups = divide_groups(arr, result, M)

    # 결과 출력
    print(result)
    print(" ".join(map(str, best_groups)))

solve()

File: 백준/Gold/2613. 숫자구슬/숫자구슬.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M;

    static int check(int[] arr, int maxSum) {
        int cnt = 1, groupSum = 0;
        for (int num : arr) {
            if (groupSum + num > maxSum) {
                cnt++;
                groupSum = num;
            } else {
                groupSum += num;
            }
        }
        return cnt;
    }

    static List<Integer> divide_groups(int[] arr, int maxSum, int M) {
        List<Integer> groups = new ArrayList<>();
        int currentSum = 0, currentCount = 0;
        int extraGroups = check(arr, maxSum) - M;
        
        for (int num : arr) {
            if (currentSum + num > maxSum || (extraGroups > 0 && groups.size() + (arr.length - groups.stream().mapToInt(Integer::intValue).sum()) > M)) {
                groups.add(currentCount);
                currentSum = num;
                currentCount = 1;
                extraGroups--;
            } else {
                currentSum += num;
                currentCount++;
            }
        }
        groups.add(currentCount);
        
        while (groups.size() < M) {
            for (int i = 0; i < groups.size(); i++) {
                if (groups.get(i) > 1) {
                    groups.set(i, groups.get(i) - 1);
                    groups.add(i + 1, 1);
                    if (groups.size() == M) break;
                }
            }
        }
        return groups;
    }

    static void solve() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        
        int[] arr = new int[N];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        
        int low = Arrays.stream(arr).max().getAsInt();
        int high = Arrays.stream(arr).sum();
        int result = high;
        
        while (low <= high) {
            int mid = (low + high) / 2;
            if (check(arr, mid) <= M) {
                result = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        
        List<Integer> bestGroups = divide_groups(arr, result, M);
        
        System.out.println(result);
        for (int i = 0; i < bestGroups.size(); i++) {
            System.out.print(bestGroups.get(i));
            if (i < bestGroups.size() - 1) System.out.print(" ");
        }
        System.out.println();
    }

    public static void main(String[] args) throws IOException {
        solve();
    }
}